src/get-accname.ts

```diff
@@ -0,0 +1,207 @@
+
+  // 4.1.9 figure Element Accessible Name Computation
+  // TODO
+
+  // 4.1.10 img Element Accessible Name Computation
+  // TODO
+
+  // 4.1.11 table Element Accessible Name Computation
+  // TODO
+
+  // 4.1.12 tr, td, th Elements Accessible Name Computation
+  // TODO
+
+  // 4.1.13 a Element Accessible Name Computation
+  // TODO
+
+  // 4.1.14 area Element Accessible Name Computation
+  // TODO
+
+  // 4.1.15 iframe Element Accessible Name Computation
+  // TODO
+
+  // 4.1.16 Section and Grouping Element Accessible Name Computation
+  // TODO
+
+  // 4.1.17 Text-level Element Accessible Name Computation
+  // TODO
+}
+
+export function getAccdescription(element: Element): string | undefined {
+  if (typeof Element === 'undefined') {
+    throw new Error('getAccdescription is not supported in this environment.');
+  }
+
+  // 1. aria-describedby
+  if (element.getAttribute('aria-describedby')) {
+    return concatIDRefList(element.getAttribute('aria-describedby')!);
+  }
+
+  // 2. aria-description
+  if (element.getAttribute('aria-description')) {
+    return element.getAttribute('aria-description')!.trim();
+  }
+
+  return undefined;
+}
+
+/**
+ * Turn an array of potential values into a concatenated string, separated by " "
+ * @see https://w3c.github.io/accname/#mapping_additional_nd_description
+ */
+function concatIDRefList(idRefList: string): string {
+  const idList = idRefList.trim().split(/\s+/);
+  const nameList: (string | undefined)[] = [];
+  for (const id of idList) {
+    nameList.push(document.getElementById(id)?.textContent || undefined);
+  }
+  // Note: this filters out empty strings (""), which are significant in some contexts, but not for naming.
+  return nameList.filter(Boolean).join(' ').trim();
+}
+
+/**
+ * Is a given element hidden?
+ * @see https://w3c.github.io/accname/#dfn-hidden
+ */
+function isHidden(element: Element): boolean {
+  // Note: <div aria-hidden> will yield "", so equate that with "true"
+  if (['true', ''].includes(element.getAttribute('aria-hidden')!)) {
+    return true;
+  }
+
+  // Check if the element is hidden by CSS
+  // https://w3c.github.io/accname/#computation-steps
+  const style = window.getComputedStyle(element);
+  return (
+    style.display === 'none' ||
+    style.visibility === 'hidden' ||
+    style.visibility === 'collapse' ||
+    style.contentVisibility === 'hidden'
+    // Note: opacity: 0 is NOT hidden
+  );
+}
+
+/** Get accessible name from associated <label>s */
+function getAccnameFromHTMLFor(element: Element): string | undefined {
+  if (!element.id) {
+    return;
+  }
+  const labels = Array.from(document.querySelectorAll(`label[for="${element.id}"]`));
+  const accList: (string | undefined)[] = [];
+  for (const label of labels) {
+    accList.push(label.textContent?.trim());
+  }
+  return accList.filter(Boolean).join(' ') || undefined;
+}
+
+/** Get text content while ignoring hidden nodes */
+function getVisibleText(element: Element, name = '') {
```

Comment:
Note: because this library is designed to work inside JSDOM, I’m at a bit of a crossroads here. In a true browser env, `.innerText` _should_ just work as expected. But only when there’s a rendering engine enabled. JSDOM will produce [different results](https://github.com/jsdom/jsdom/issues/1245), which is known and expected. But even given this, I’m not sure how to test for this without it being a proxy for something else, but I’d be keen to do `.innerText` in browsers, and a custom implementation in JSDOM.